
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://xin-huang.github.io/sai/userguide/score/">
      
      
        <link rel="prev" href="../..">
      
      
        <link rel="next" href="../outlier/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.40">
    
    
      
        <title>Calculate Statistics - sai</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.8c3ca2c6.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#calculate-statistics" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="sai" class="md-header__button md-logo" aria-label="sai" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            sai
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Calculate Statistics
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/xin-huang/sai" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="sai" class="md-nav__button md-logo" aria-label="sai" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    sai
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/xin-huang/sai" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    User Guide
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            User Guide
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Calculate Statistics
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Calculate Statistics
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#input-files" class="md-nav__link">
    <span class="md-ellipsis">
      Input files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output-files" class="md-nav__link">
    <span class="md-ellipsis">
      Output files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#examples" class="md-nav__link">
    <span class="md-ellipsis">
      Examples
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../outlier/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Extract Outliers
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#input-files" class="md-nav__link">
    <span class="md-ellipsis">
      Input files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output-files" class="md-nav__link">
    <span class="md-ellipsis">
      Output files
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#examples" class="md-nav__link">
    <span class="md-ellipsis">
      Examples
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="calculate-statistics">Calculate Statistics</h1>
<p><code>sai</code> provides the <code>score</code> command to calculate different statistics.</p>
<p>To see available options, we can use the following command:</p>
<div class="codehilite"><pre><span></span><code>sai score -h
</code></pre></div>

<p>This will show information for each argument:</p>
<table>
<thead>
<tr>
<th>Argument</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>-h, --help</td>
<td>show this help message and exit</td>
</tr>
<tr>
<td>--vcf</td>
<td>Path to the VCF file containing variant data</td>
</tr>
<tr>
<td>--chr-name</td>
<td>Chromosome name to analyze from the VCF file.</td>
</tr>
<tr>
<td>--win-len</td>
<td>Length of each genomic window in base pairs. Default: 50,000.</td>
</tr>
<tr>
<td>--win-step</td>
<td>Step size in base pairs between consecutive windows. Default: 10,000.</td>
</tr>
<tr>
<td>--anc-alleles</td>
<td>Path to the BED file with ancestral allele information. If ancestral allele information is not provided, filtering will be performed for each variant based on whether the allele frequency of any allele (assuming biallelic) meets the specified condition during the calculation of the statistics. Default: None.</td>
</tr>
<tr>
<td>--output</td>
<td>Output file path for saving results.</td>
</tr>
<tr>
<td>--config</td>
<td>Path to the YAML configuration file specifying the statistics to compute, ploidy settings, and population group file paths.</td>
</tr>
</tbody>
</table>
<h2 id="input-files">Input files</h2>
<p><code>sai</code> requires a biallelic VCF file containing all populations, including the reference population (assumed to have no introgression), the target population (assumed to have received introgressed material), and the source population (assumed to have provided the introgressed material). It also assumes that all variants observed in these populations are already included in the input VCF file. The program does not fill in missing genotypes (e.g., treating unobserved genotypes as homozygous reference, as done in <a href="https://github.com/xzhang-popgen/maladapt/blob/fba58dac6278d1b4cc573c78ab7c56911b4072b0/empirical/1empirical_compute-features.py#L318-L326">MaLAdapt</a>). An example VCF file can be found <a href="https://github.com/xin-huang/sai/blob/main/examples/data/1KG.nea_den.chr9.example.vcf.gz">here</a>.</p>
<p>Three tab-delimited files are required to specify the population information for individuals in the reference, target, and source populations. An example file is shown below and can also be found <a href="https://github.com/xin-huang/sai/blob/main/examples/data/1KG.nea_den.samples.txt">here</a>.</p>
<div class="codehilite"><pre><span></span><code>NEA AltaiNeandertal
DEN Denisova
</code></pre></div>

<p>This file specifies the individuals in the source populations. In this example, there are two source populations, NEA and DEN (first column), and the second column lists the corresponding sample names matching those in the VCF header.</p>
<p>A configuration file in <a href="https://yaml.org/">YAML format</a> is required to specify the statistics to compute, ploidy settings, and population group file paths. An example can be found <a href="https://github.com/xin-huang/sai/blob/main/examples/data/1KG.nea_den.chr9.example.both.config.yaml">here</a>. In the example below, the <code>statistics</code> section lists which statistics will be computed. Empty braces (<code>{}</code>) indicate that no additional parameters are required for that statistic, while entries such as $U$ and $Q$ specify reference (<code>ref</code>), target (<code>tgt</code>), and source (<code>src</code>) populations with filtering conditions. The <code>ploidies</code> section defines the ploidy level for each population, typically 2 for diploids. The populations section provides file paths to sample lists that assign individuals to each population.</p>
<div class="codehilite"><pre><span></span><code><span class="n">statistics</span><span class="o">:</span>
<span class="w">  </span><span class="n">Danc</span><span class="o">:</span><span class="w"> </span><span class="o">{}</span>
<span class="w">  </span><span class="n">DD</span><span class="o">:</span><span class="w"> </span><span class="o">{}</span>
<span class="w">  </span><span class="n">Dplus</span><span class="o">:</span><span class="w"> </span><span class="o">{}</span>
<span class="w">  </span><span class="n">df</span><span class="o">:</span><span class="w"> </span><span class="o">{}</span>
<span class="w">  </span><span class="n">fd</span><span class="o">:</span><span class="w"> </span><span class="o">{}</span>
<span class="w">  </span><span class="n">U</span><span class="o">:</span>
<span class="w">    </span><span class="n">ref</span><span class="o">:</span><span class="w"> </span>
<span class="w">      </span><span class="n">AFR</span><span class="o">+</span><span class="n">EAS</span><span class="o">:</span><span class="w"> </span><span class="mf">0.01</span>
<span class="w">    </span><span class="n">tgt</span><span class="o">:</span><span class="w"> </span>
<span class="w">      </span><span class="n">EUR</span><span class="o">:</span><span class="w"> </span><span class="mf">0.5</span>
<span class="w">    </span><span class="n">src</span><span class="o">:</span>
<span class="w">      </span><span class="n">NEA</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;=1&quot;</span>
<span class="w">      </span><span class="n">DEN</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;=1&quot;</span>
<span class="w">  </span><span class="n">Q</span><span class="o">:</span>
<span class="w">    </span><span class="n">ref</span><span class="o">:</span><span class="w"> </span>
<span class="w">      </span><span class="n">AFR</span><span class="o">+</span><span class="n">EAS</span><span class="o">:</span><span class="w"> </span><span class="mf">0.01</span>
<span class="w">    </span><span class="n">tgt</span><span class="o">:</span><span class="w"> </span>
<span class="w">      </span><span class="n">EUR</span><span class="o">:</span><span class="w"> </span><span class="mf">0.95</span>
<span class="w">    </span><span class="n">src</span><span class="o">:</span>
<span class="w">      </span><span class="n">NEA</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;=1&quot;</span>
<span class="w">      </span><span class="n">DEN</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;=1&quot;</span>

<span class="n">ploidies</span><span class="o">:</span>
<span class="w">  </span><span class="n">ref</span><span class="o">:</span>
<span class="w">    </span><span class="n">AFR</span><span class="o">+</span><span class="n">EAS</span><span class="o">:</span><span class="w"> </span><span class="mi">2</span>
<span class="w">  </span><span class="n">tgt</span><span class="o">:</span>
<span class="w">    </span><span class="n">EUR</span><span class="o">:</span><span class="w"> </span><span class="mi">2</span>
<span class="w">  </span><span class="n">src</span><span class="o">:</span>
<span class="w">    </span><span class="n">NEA</span><span class="o">:</span><span class="w"> </span><span class="mi">2</span>
<span class="w">    </span><span class="n">DEN</span><span class="o">:</span><span class="w"> </span><span class="mi">2</span>

<span class="n">populations</span><span class="o">:</span>
<span class="w">  </span><span class="n">ref</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;examples/data/1KG.ref.samples.txt&quot;</span>
<span class="w">  </span><span class="n">tgt</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;examples/data/1KG.tgt.samples.txt&quot;</span>
<span class="w">  </span><span class="n">src</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;examples/data/1KG.nea_den.samples.txt&quot;</span>
</code></pre></div>

<h2 id="output-files">Output files</h2>
<p>An example output is shown below.</p>
<table>
<thead>
<tr>
<th>Chrom</th>
<th>Start</th>
<th>End</th>
<th>Ref</th>
<th>Tgt</th>
<th>Src</th>
<th>N(Variants)</th>
<th>Danc.NEA</th>
<th>Danc.DEN</th>
<th>DD.NEA</th>
<th>DD.DEN</th>
<th>Dplus.NEA</th>
<th>Dplus.DEN</th>
<th>df.NEA</th>
<th>df.DEN</th>
<th>fd.NEA</th>
<th>fd.DEN</th>
<th>U</th>
<th>Q</th>
</tr>
</thead>
<tbody>
<tr>
<td>9</td>
<td>16400001</td>
<td>16440000</td>
<td>AFR+EAS</td>
<td>EUR</td>
<td>NEA,DEN</td>
<td>1082</td>
<td>-0.12355029565477203</td>
<td>-0.09559566661094993</td>
<td>-7.478853072043918</td>
<td>-4.739448231247437</td>
<td>-0.14512264357342178</td>
<td>-0.09212831462235394</td>
<td>-0.021122428026740657</td>
<td>-0.005965214141977462</td>
<td>-0.03412663364767112</td>
<td>-0.011254553402083362</td>
<td>0</td>
<td>nan</td>
</tr>
<tr>
<td>9</td>
<td>16440001</td>
<td>16480000</td>
<td>AFR+EAS</td>
<td>EUR</td>
<td>NEA,DEN</td>
<td>983</td>
<td>0.23853287087522393</td>
<td>0.09720568452718485</td>
<td>15.824720328820732</td>
<td>2.6519198302249976</td>
<td>0.24254139814192216</td>
<td>0.04198624490861517</td>
<td>0.06679895300702358</td>
<td>-0.027612395797831944</td>
<td>0.09662599519731511</td>
<td>-0.06072062935032934</td>
<td>0</td>
<td>nan</td>
</tr>
<tr>
<td>9</td>
<td>16480001</td>
<td>16520000</td>
<td>AFR+EAS</td>
<td>EUR</td>
<td>NEA,DEN</td>
<td>985</td>
<td>0.11191758509638665</td>
<td>0.1371820707366662</td>
<td>4.845944176212569</td>
<td>6.488856543280008</td>
<td>0.08062408365129729</td>
<td>0.10781131967371557</td>
<td>-0.0059518440634471394</td>
<td>0.006339448215182395</td>
<td>-0.009777268672625093</td>
<td>0.011579864693126562</td>
<td>0</td>
<td>nan</td>
</tr>
<tr>
<td>9</td>
<td>16520001</td>
<td>16560000</td>
<td>AFR+EAS</td>
<td>EUR</td>
<td>NEA,DEN</td>
<td>966</td>
<td>0.008890710141582193</td>
<td>0.0017874385890340458</td>
<td>-3.2895208116381127</td>
<td>-3.762141437091735</td>
<td>-0.07409036881532824</td>
<td>-0.08473527378923923</td>
<td>-0.08168801665077327</td>
<td>-0.0882036818431493</td>
<td>-0.10989219575082872</td>
<td>-0.11581875355584634</td>
<td>0</td>
<td>nan</td>
</tr>
<tr>
<td>9</td>
<td>16560001</td>
<td>16600000</td>
<td>AFR+EAS</td>
<td>EUR</td>
<td>NEA,DEN</td>
<td>1029</td>
<td>0.0024719610094924406</td>
<td>0.013019347645432937</td>
<td>-3.0949185837356765</td>
<td>-2.3295899207926993</td>
<td>-0.07022371194458008</td>
<td>-0.05285840228768851</td>
<td>-0.024174513563260166</td>
<td>-0.021736551375934953</td>
<td>-0.05323265434587958</td>
<td>-0.051743268070762606</td>
<td>0</td>
<td>0.0008946322067594431</td>
</tr>
<tr>
<td>9</td>
<td>16600001</td>
<td>16640000</td>
<td>AFR+EAS</td>
<td>EUR</td>
<td>NEA,DEN</td>
<td>1063</td>
<td>0.022253891584592442</td>
<td>-0.05846778040121223</td>
<td>3.258157404777677</td>
<td>-3.199939253368683</td>
<td>0.07998846049314343</td>
<td>-0.06970294049702096</td>
<td>0.025940599557366877</td>
<td>-0.0056209360304606635</td>
<td>0.10418231535859726</td>
<td>-0.01922836158621726</td>
<td>0</td>
<td>0.0</td>
</tr>
<tr>
<td>9</td>
<td>16640001</td>
<td>16680000</td>
<td>AFR+EAS</td>
<td>EUR</td>
<td>NEA,DEN</td>
<td>1153</td>
<td>0.10931478521694758</td>
<td>-0.02727035560668536</td>
<td>11.405249061188428</td>
<td>-2.9954578087033354</td>
<td>0.17733519261177239</td>
<td>-0.04657505369825686</td>
<td>0.07092801200718006</td>
<td>-0.01271958977736748</td>
<td>0.14831905506347323</td>
<td>-0.0603694010038118</td>
<td>0</td>
<td>nan</td>
</tr>
<tr>
<td>9</td>
<td>16680001</td>
<td>16720000</td>
<td>AFR+EAS</td>
<td>EUR</td>
<td>NEA,DEN</td>
<td>999</td>
<td>0.15113133237922183</td>
<td>0.14710421254212552</td>
<td>18.309291078923295</td>
<td>17.05495597841523</td>
<td>0.25761194634556367</td>
<td>0.24391495232933705</td>
<td>0.1898707100981595</td>
<td>0.17809065689988668</td>
<td>0.22245899486649723</td>
<td>0.3768632297159813</td>
<td>1</td>
<td>0.6282306163021869</td>
</tr>
<tr>
<td>9</td>
<td>16720001</td>
<td>16760000</td>
<td>AFR+EAS</td>
<td>EUR</td>
<td>NEA,DEN</td>
<td>1007</td>
<td>0.4129294073240463</td>
<td>0.20572780178794584</td>
<td>53.56721772539367</td>
<td>29.49165522736596</td>
<td>0.49454474017007327</td>
<td>0.2692256133787326</td>
<td>0.45835432326244324</td>
<td>0.2539470613190382</td>
<td>0.5958374858338679</td>
<td>0.354370074530527</td>
<td>2</td>
<td>0.6828528827037773</td>
</tr>
<tr>
<td>9</td>
<td>16760001</td>
<td>16800000</td>
<td>AFR+EAS</td>
<td>EUR</td>
<td>NEA,DEN</td>
<td>1060</td>
<td>0.40683688578049354</td>
<td>0.1438067334121798</td>
<td>40.19722340559816</td>
<td>16.581497128341063</td>
<td>0.5263549295804029</td>
<td>0.2153953867869153</td>
<td>0.5848435316604244</td>
<td>0.21982990433919478</td>
<td>0.5444309862037052</td>
<td>0.23506396309346742</td>
<td>1</td>
<td>0.6988071570576541</td>
</tr>
<tr>
<td>9</td>
<td>16800001</td>
<td>16840000</td>
<td>AFR+EAS</td>
<td>EUR</td>
<td>NEA,DEN</td>
<td>960</td>
<td>0.40565740562034003</td>
<td>0.22930692442169218</td>
<td>24.16718735207801</td>
<td>13.914000126226764</td>
<td>0.5158532156040262</td>
<td>0.29611831851629133</td>
<td>0.3520651233378468</td>
<td>0.1984114833463481</td>
<td>0.6411180111472388</td>
<td>0.5318202119747509</td>
<td>1</td>
<td>0.7176938369781312</td>
</tr>
<tr>
<td>9</td>
<td>16840001</td>
<td>16880000</td>
<td>AFR+EAS</td>
<td>EUR</td>
<td>NEA,DEN</td>
<td>923</td>
<td>0.21619895332973812</td>
<td>0.19275643241641574</td>
<td>12.284513159140396</td>
<td>10.489264807977534</td>
<td>0.22271722145226366</td>
<td>0.19016951529508055</td>
<td>0.09059952770351988</td>
<td>0.058657408877032656</td>
<td>0.08230019337435454</td>
<td>0.06443318954095166</td>
<td>0</td>
<td>0.15382703777335985</td>
</tr>
<tr>
<td>9</td>
<td>16880001</td>
<td>16920000</td>
<td>AFR+EAS</td>
<td>EUR</td>
<td>NEA,DEN</td>
<td>512</td>
<td>0.3881193782134191</td>
<td>0.3589872405156586</td>
<td>20.739761825870175</td>
<td>19.506952333617342</td>
<td>0.48085645624316853</td>
<td>0.4522734663012016</td>
<td>0.5199698990819357</td>
<td>0.5044977777777778</td>
<td>0.47736503902045885</td>
<td>0.55657933398083</td>
<td>0</td>
<td>nan</td>
</tr>
</tbody>
</table>
<p>Each row in the output file corresponds to a genomic window, summarizing the result for that region. The columns are as follows:</p>
<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>Chrom</td>
<td>Chromosome number where the window is located.</td>
</tr>
<tr>
<td>Start</td>
<td>Start position (1-based, inclusive) of the window.</td>
</tr>
<tr>
<td>End</td>
<td>End position (inclusive) of the window.</td>
</tr>
<tr>
<td>Ref</td>
<td>Reference population.</td>
</tr>
<tr>
<td>Tgt</td>
<td>Target population.</td>
</tr>
<tr>
<td>Src</td>
<td>Source population.</td>
</tr>
<tr>
<td>N(Variants)</td>
<td>Number of variants found in this genomic window. Sites with missing genotypes in any population are excluded. If ancestral allele information is provided, variants whose ancestral allele differ from both the reference and alternative allele in the VCF file are also excluded.</td>
</tr>
<tr>
<td>Danc.src1</td>
<td>Value of the $D_{anc}$ statistic using src1 as the source population</td>
</tr>
<tr>
<td>Danc.src2</td>
<td>Value of the $D_{anc}$ statistic using src2 as the source population</td>
</tr>
<tr>
<td>DD.src1</td>
<td>Value of the $D_D$ statistic using src1 as the source population</td>
</tr>
<tr>
<td>DD.src2</td>
<td>Value of the $D_D$ statistic using src2 as the source population</td>
</tr>
<tr>
<td>Dplus.src1</td>
<td>Value of the $D^+$ statistic using src1 as the source population</td>
</tr>
<tr>
<td>Dplus.src2</td>
<td>Value of the $D^+$ statistic using src2 as the source population</td>
</tr>
<tr>
<td>df.src1</td>
<td>Value of the $d_f$ statistic using src1 as the source population</td>
</tr>
<tr>
<td>df.src2</td>
<td>Value of the $d_f$ statistic using src2 as the source population</td>
</tr>
<tr>
<td>fd.src1</td>
<td>Value of the $f_d$ statistic using src1 as the source population</td>
</tr>
<tr>
<td>fd.src2</td>
<td>Value of the $f_d$ statistic using src2 as the source population</td>
</tr>
<tr>
<td>U</td>
<td>Value of the $U$ statistic using both src1 and src2 as the source populations</td>
</tr>
<tr>
<td>Q</td>
<td>Value of the $Q$ statistic using both src1 and src2 as the source populations</td>
</tr>
</tbody>
</table>
<h2 id="examples">Examples</h2>
<p>The following example estimates all the statistics defined by the example configuration above with biallelic single nucleotide polymorphisms (SNPs) in the region <code>chr9:16400001-16900000</code> from <a href="https://ftp.ncbi.nih.gov/1000genomes/ftp/release/20130502/">the 1000 Genomes Project</a>. The Neanderthal genome was obtained from <a href="http://cdna.eva.mpg.de/neandertal/Vindija/VCF/Altai/">here</a>, and the Denisovan genome from <a href="http://cdna.eva.mpg.de/neandertal/Vindija/VCF/Denisova/">here</a>. The reference population includes all African populations (ESN, GWD, LWK, MSL, YRI), excluding the admixed African populations (ACB and ASW), and all East Asian populations (CDX, CHB, CHS, JPT, KHV). The target population consists of all European populations (CEU, FIN, GBR, IBS, TSI). The source populations comprise one Neanderthal individual (AltaiNeandertal) and one Denisovan individual (Denisova).</p>
<p>To detect adaptively introgressed variants from <strong>both Neanderthals and Denisovans</strong>, we can use the following command:</p>
<div class="codehilite"><pre><span></span><code>sai score --vcf examples/data/1KG.nea_den.chr9.example.vcf.gz \
          --anc-alleles examples/data/hg19.chr9.example.anc.alleles.bed \
          --chr-name 9 \
          --win-len 40000 --win-step 40000 \
          --config examples/data/1KG.nea_den.chr9.example.both.config.yaml \
          --output examples/results/both/1KG.nea_den.chr9.example.both.stats.scores.tsv
</code></pre></div>

<p>We provide a BED file to specify the ancestral allele for each variant, based on alignments from <a href="https://ftp.ensembl.org/pub/release-75/fasta/ancestral_alleles/">Ensembl</a>.
In this example, for the $U$ statistic, the <code>ref</code> entry in the configuration file specifies that the allele frequency must be less than 0.01 in the AFR+EAS population, the <code>src</code> entries require the allele to be fixed (<code>=1</code>) in both Neanderthal and Denisovan genomes, and the <code>tgt</code> entry indicates that the allele frequency must be greater than 0.5 in the EUR population. Similarly, for the $Q$ statistic, the <code>tgt</code> entry specifies the quantile of the derived allele frequencies in the EUR population at 0.95. The output is shown above and can be found <a href="https://github.com/xin-huang/sai/blob/main/examples/results/both/1KG.nea_den.chr9.example.both.stats.scores.tsv">here</a>.</p>
<p>The <code>src</code> entries can be set as Nea: "=1", Den: "=0" or the reverse to identify lineage-specific variants:</p>
<div class="codehilite"><pre><span></span><code><span class="n">statistics</span><span class="o">:</span>
<span class="w">  </span><span class="n">U</span><span class="o">:</span>
<span class="w">    </span><span class="n">ref</span><span class="o">:</span><span class="w"> </span>
<span class="w">      </span><span class="n">AFR</span><span class="o">+</span><span class="n">EAS</span><span class="o">:</span><span class="w"> </span><span class="mf">0.01</span>
<span class="w">    </span><span class="n">tgt</span><span class="o">:</span><span class="w"> </span>
<span class="w">      </span><span class="n">EUR</span><span class="o">:</span><span class="w"> </span><span class="mf">0.5</span>
<span class="w">    </span><span class="n">src</span><span class="o">:</span>
<span class="w">      </span><span class="n">NEA</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;=1&quot;</span>
<span class="w">      </span><span class="n">DEN</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;=0&quot;</span>
<span class="w">  </span><span class="n">Q</span><span class="o">:</span>
<span class="w">    </span><span class="n">ref</span><span class="o">:</span><span class="w"> </span>
<span class="w">      </span><span class="n">AFR</span><span class="o">+</span><span class="n">EAS</span><span class="o">:</span><span class="w"> </span><span class="mf">0.01</span>
<span class="w">    </span><span class="n">tgt</span><span class="o">:</span><span class="w"> </span>
<span class="w">      </span><span class="n">EUR</span><span class="o">:</span><span class="w"> </span><span class="mf">0.95</span>
<span class="w">    </span><span class="n">src</span><span class="o">:</span>
<span class="w">      </span><span class="n">NEA</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;=1&quot;</span>
<span class="w">      </span><span class="n">DEN</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;=0&quot;</span>
</code></pre></div>

<p>An example configuration file for Neanderthal-specific introgressed variants can be found <a href="https://github.com/xin-huang/sai/blob/main/examples/data/1KG.nea_den.chr9.example.nea.specific.config.yaml">here</a>, with its corresponding output available <a href="https://github.com/xin-huang/sai/blob/main/examples/results/nea_specific/1KG.nea_den.chr9.example.nea.specific.stats.scores.tsv">here</a>. Similarly, an example configuration file for Denisovan-specific introgressed variants is provided <a href="https://github.com/xin-huang/sai/blob/main/examples/data/1KG.nea_den.chr9.example.den.specific.config.yaml">here</a>, with the output available <a href="https://github.com/xin-huang/sai/blob/main/examples/results/den_specific/1KG.nea_den.chr9.example.den.specific.stats.scores.tsv">here</a>. Note that only the $U$ and $Q$ statistics support multi-source introgression filtering.</p>
<p>In addition, log files corresponding to the $U$ or $Q$ statistics are generated to record the positions of variants used in their calculation. An example is shown below:</p>
<table>
<thead>
<tr>
<th>Chrom</th>
<th>Start</th>
<th>End</th>
<th>U_SNP</th>
</tr>
</thead>
<tbody>
<tr>
<td>9</td>
<td>16400001</td>
<td>16440000</td>
<td>NA</td>
</tr>
<tr>
<td>9</td>
<td>16440001</td>
<td>16480000</td>
<td>NA</td>
</tr>
<tr>
<td>9</td>
<td>16480001</td>
<td>16520000</td>
<td>NA</td>
</tr>
<tr>
<td>9</td>
<td>16520001</td>
<td>16560000</td>
<td>NA</td>
</tr>
<tr>
<td>9</td>
<td>16560001</td>
<td>16600000</td>
<td>NA</td>
</tr>
<tr>
<td>9</td>
<td>16600001</td>
<td>16640000</td>
<td>NA</td>
</tr>
<tr>
<td>9</td>
<td>16640001</td>
<td>16680000</td>
<td>NA</td>
</tr>
<tr>
<td>9</td>
<td>16680001</td>
<td>16720000</td>
<td>9:16715657</td>
</tr>
<tr>
<td>9</td>
<td>16720001</td>
<td>16760000</td>
<td>9:16758710,9:16759357</td>
</tr>
<tr>
<td>9</td>
<td>16760001</td>
<td>16800000</td>
<td>9:16770229</td>
</tr>
<tr>
<td>9</td>
<td>16800001</td>
<td>16840000</td>
<td>9:16800789</td>
</tr>
<tr>
<td>9</td>
<td>16840001</td>
<td>16880000</td>
<td>NA</td>
</tr>
<tr>
<td>9</td>
<td>16880001</td>
<td>16920000</td>
<td>NA</td>
</tr>
</tbody>
</table>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": [], "search": "../../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.525ec568.min.js"></script>
      
        <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML"></script>
      
    
  </body>
</html>